lib/sharing/exercise_metadata.dart

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// An exception thrown when a metadata file has missing or invalid fields.
+class MetadataException implements Exception {
+  final String message;
+
+  const MetadataException(this.message);
+}
+
+/// Modes for individual exercises.
+///
+/// The DartPad editors, output windows, and so on will be configured according
+/// to this value.
+enum ExerciseMode {
+  dart,
+  html,
+  flutter,
+}
+
+final exerciseModeNames = <String, ExerciseMode>{
+  'dart': ExerciseMode.dart,
+  'html': ExerciseMode.html,
+  'flutter': ExerciseMode.flutter,
+};
+
+/// Metadata for a single file within a larger exercise.
+class ExerciseFileMetadata {
+  String name;
+  String alternatePath;
+
+  String get path =>
+      (alternatePath == null || alternatePath.isEmpty) ? name : alternatePath;
+
+  ExerciseFileMetadata.fromJson(Map<String, dynamic> json) {
+    if (json == null) {
+      throw MetadataException('Null json was given to ExerciseFileMetadata().');
+    }
+
+    if (json['name'] == null ||
+        json['name'] is! String ||
+        json['name'].isEmpty) {
+      throw MetadataException('The "name" field is required for each file.');
+    }
+
+    name = json['name'];
+    alternatePath = json['alternatePath'];
+  }
+}
+
+/// Represents the metadata for a single codelab exercise, as defined by the
+/// exercise's `dartpad-metadata.json` file.
```

Comment:
nit: I would prefer to stick with underscores since the Dart file convention is to use underscores

Reply:
Good nit. I've revised.